Adds e2e firewall tests for LoadBalancer service, ingress, and e2e cluster

MrHohn opened the pull request:
Fixes #25488 and fixes #31827.

This PR adds e2e firewall test for LoadBalancer type service, ingress and e2e cluster.

Test details for LoadBalancer type service as below:
- Verifies corresponding firewall rule has correct `sourceRanges`, `ports and protocols` and `target tags`.
- Verifies requests can reach all expected instances.
- Verifies requests can not reach instances that are not included.

Overview of the test procedure:
- Creates a LoadBalancer type service.
- Validates the corresponding firewall rule.
- Creates netexec pods as service backends.
- Sends requests from outside of the cluster and examine hitting all instances in range.
- Removes tags from one of the instances in order to get it out of firewall rule's range.
- Sends requests from outside of the cluster and examine not hitting this instance.
- Recovers tags for this instances and verifies its traffic is back.

@bprashanth @bowei @thockin

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36546)
<!-- Reviewable:end -->

MrHohn commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,411 @@
+			"error waiting for daemon pod to start")
+		Expect(checkDaemonStatus(f, dsName)).NotTo(HaveOccurred())
+
+		// Send request from outside of the cluster because internal pods trying to reach external
```
I realize this comment is not clear. Accessing from outside because e2e cluster has this rule:
```
e2e-default-internal-node 10.0.0.0/8 tcp:1-65535, udp:1-65535, icmp e2e-test-zihongz-minion
```
And it will allow internal traffic to reach any nodes.

MrHohn commented:
@k8s-bot unit test this, issue: #36473

MrHohn commented:
@k8s-bot cvm gke e2e test this, issue: #33388

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+		serviceName := "firewall-test-loadbalancer"
+		jig := NewServiceTestJig(c, serviceName)
+		svc := createOnlyLocalLoadBalancServiceWithPort(jig, ns, serviceName, firewallTestHttpPort, firewallTestSourceRanges)
+		defer func() {
```
you want to change the type to !LoadBalancer before deleting it, otherwise if this is the last test cluster-teardown will shutdown the master and leak your lb resources

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+
+		// Send requests from outside of the cluster because internal traffic is whitelisted
+		By("Accessing the external service ip from outside, all non-master nodes should be reached")
+		nodeNames, err := getNonMaterNodeNames(c)
```
s/Mater/Master

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+}
+
+func sameStringElements(result, expected []string) error {
+	ex := make(map[string]bool)
```
can you just use sets.NewSet to compare? you can basically parse the list into a Set directly, and then use Equals/Diff etc: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/networking_utils.go#L236

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+
+func getNonMaterNodeNames(c clientset.Interface) (sets.String, error) {
+	masterName := framework.TestContext.CloudConfig.MasterName
+	nodeList, err := c.Core().Nodes().List(api.ListOptions{})
```
I don't think you want to get all nodes, see how the services test handles this (limits nodes in large clusters, because that will take hours to complete).

MrHohn replied:
Changes to use `maxNodesForEndpointsTests` pods.

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+		Timeout:   timeout,
+		Transport: &http.Transport{},
+	}
+	url := fmt.Sprintf("http://%v:%v/%v", ip, port, path)
```
there's probably already a function that does this (maybe it's called simpleGET?). If we don't have such a function, suggest adding it to framework since it's really common.

bprashanth replied:
Just make it take the entier url, vs ip, path, port seperately

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+			return false, nil
+		}
+		tagsFromGCE = res.Tags.Items
+		if err := sameStringElements(tagsFromGCE, tags); err != nil {
```
sameStringElements is a slightly confusing name, suggest a rename + docstring to explain what the purpose of the function is

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+
+	framework.Logf("Waiting up to %v for not hitting %v for %v times", timeout, targetHost, countToSucceed)
+	condition := func() (bool, error) {
+		respBody, err := queryFromUrl(externalIP, "hostname", httpPort, firewallTestHttpTimeout)
```
I think we're alreayd doing this in various places (hitting externalip:nodePort/url from a jenkins builder). Can you either use the same pattern as those places or try to centralize the logic for reuse? eg: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/service.go#L1242

MrHohn replied:
Using `testReachableHTTPWithContentTimeout()` this time.

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+	return nodeNames, nil
+}
+
+func examineHitNodesFromOutside(externalIP string, httpPort int32, timeout time.Duration, expectedHosts sets.String, exitWhenSatisfy bool) error {
```
can you try to combine this and the next function since they're so similar?

MrHohn replied:
I found it is actually not easy to combine them together considering numbers of different corner cases. I prefer to keep them separate unless I have a good idea.

MrHohn replied:
These two functions are combined.

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+	return fmt.Sprintf("k8s-fw-%s", name)
+}
+
+func getNodesTagFromGroupName(name string) (string, error) {
```
can you get the tags from the instance template instaed? less fickle than parsing it out of the group name, because even if someone changes the scheme for allocating node tags, it will always be correct in the instance template.

MrHohn replied:
Changes to retrieve tags from instance instead.

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+}
+
+func makeFirewallName(name string) string {
+	return fmt.Sprintf("k8s-fw-%s", name)
```
please add a comment saying that this should match the formatting in cloudprovider/gce/gce.go or wherever, so when it breaks, please will know where to look to find the thing they should be correcting it to.

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+	if err := verifySourceRanges(fw.SourceRanges, svc.Spec.LoadBalancerSourceRanges); err != nil {
+		return err
+	}
+	framework.Logf("Source ranges are correct.")
```
you can avoid logging anything on success, and just on failure, since it just adds noise when debugging some other failure later in the test

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+		Expect(err).NotTo(HaveOccurred())
+		Expect(examineHitNodesFromOutside(svcExternalIP, firewallTestHttpPort, firewallTimeoutDefault, nodeNames, true)).NotTo(HaveOccurred())
+
+		By("Removing tags from one of the nodes")
```
please add a comment here about exactly what we're trying to verify. Something like
```
check if there are overlapping tags on the firewall that extend beyond just the vms in our cluster, 
by removing the tag on one vm and make sure it doesn't get any traffic. This is an imperfect 
simulation, we really want to check that traffic doesn't reach a vm outside the GKE cluster, but 
that's much harder to do in the current e2e framework. 
```

bprashanth commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,410 @@
+})
+
+func createOnlyLocalLoadBalancServiceWithPort(jig *ServiceTestJig, ns, serviceName string, port int32, sourceRanges []string) *api.Service {
+	svc := jig.CreateTCPServiceOrFail(ns, func(svc *api.Service) {
```
can you reuse something from the services test by moving it into framework/services_utils or someplace? less code to maintain/update/debug when something changes globally.

bprashanth reviewed (commented):
overall looks good, nits and comments about code reuse since that will improve general debuggability and enable us to centralize log lines/tracing etc

MrHohn commented:
Thanks for reviewing this!

I plan to add one more separate general firewall e2e test here to check the `default` and `e2e` firewall setups. Will follow the suggestions and re-push a new commit latter.

MrHohn commented on test/e2e/firewall.go:
```diff
@@ -0,0 +1,471 @@
+// From cluster/gce/config-test.sh, cluster ip range is set up using below command:
+// CLUSTER_IP_RANGE="${CLUSTER_IP_RANGE:-10.180.0.0/14}"
+func getClusterIpRange() string {
+	return "10.180.0.0/14"
```
I think it is dangerous to hardcode this range. But I have not figured out a good approach to retrieve it properly. Advise needed.

MrHohn commented:
New commits are pushed. The first commit is from #36788. It fixes the change service type issue.

One more e2e test for the general e2e cluster firewall rules setup is added. It basically examines if rules are correct and confirms some well known ports are not reachable from outside.

I'm still working on moving out some common functions from service.go to service_util.go. But prefer to do that in another PR to keep this one easier to review.

MrHohn commented:
@freehan might need your advise as well.

MrHohn committed:
Adds e2e firewall tests.

For LoadBalancer type service:
- Verifies corresponding firewall rule has correct sourceRanges, ports
  & protocols, target tags.
- Verifies requests can reach all expected instances.
- Verifies requests can not reach instances that are not included.

For Ingress resrouce:
- Verifies the ingress firewall rule has correct sourceRanges, target
  tags and tcp ports.

For general e2e cluster:
- Verifies all required firewall rules has correct sourceRange, ports
  & protocols, source tags and target tags.
- Verifies well know ports on master and nodes are not
  exposed externally

MrHohn commented:
New update: a firewall test for ingress resource is also included now. This test is injected into one of the existing ingress tests. It simply verifies the ingress firewall rule has correct setup on GCE.

One assumption it makes is that the L7 default backend service's nodePort will be included in the firewall rule, even if no ingress resource actually use it as backend.

@bprashanth @freehan

bprashanth commented on test/e2e/ingress.go:
```diff
@@ -135,6 +143,32 @@ var _ = framework.KubeDescribe("Loadbalancing: L7 [Feature:Ingress]", func() {
 			ExpectNoError(pollURL(fmt.Sprintf("http://%v/", ip), "", lbPollTimeout, httpClient, true))
 
+			By("should have correct firewall rule for ingress")
+			fw, err := getFwForIngress(cloudConfig)
```
where is this implemented?

bprashanth commented on test/e2e/ingress.go:
```diff
@@ -135,6 +143,32 @@ var _ = framework.KubeDescribe("Loadbalancing: L7 [Feature:Ingress]", func() {
+			// Getting ingress backend nodePorts. Current GCE ingress controller allows
+			// traffic to the default backend nodePort by default, so retrieve that as well.
+			nodePorts := []string{}
+			defaultSvc, err := jig.client.Core().Services(api.NamespaceSystem).Get(defaultBackendName)
```
can you please make this a method on the jig called `getIngressNodePorts(ingName, ns)` that does the following:
1. get the ingress
2. get its services:service-port
3. get each service and find the nodeport associated with service-port
4. return a list of int nodeports

you can just have this method Failf internally instead of returning an error, keeps the main test cleaner and easier to reason about

MrHohn replied:
Done. No hardcoded backend name this time.

bprashanth commented on test/e2e/ingress.go:
```diff
@@ -135,6 +143,32 @@ var _ = framework.KubeDescribe("Loadbalancing: L7 [Feature:Ingress]", func() {
+			// Passed the last argument as true to verify the backend ports are
+			// included but not exactly matched, given there may be other existing
+			// ingress resources and backends we are not aware of.
+			Expect(verifyFirewallRule(fw, expFw, cloudConfig.Network, true)).NotTo(HaveOccurred())
```
I think you're missing a file in your commit? since I can't seem to find some methods.
Suggest having this as a method on the gceController like
```
func (g *gceController) verifyFirewall(ingName string, expectedTags []string{}, expectedPorts []int{}) {

    // recontruct the firewall name from ing name like https://github.com/kubernetes/kubernetes/blob/master/test/e2e/ingress_utils.go#L601
    // get the firewall object
    // insert all its ports into a sets.String
    // insert the array of given ports into a sets.String
    // check IsSuperSet
    // verify tags

}
```

bprashanth replied:
So the test might become something like
```go
expectedTags := jig.getNodeTags()
expectedPorts := jig.getIngressNodePorts(ingName, ns)
gceController.verifyFirewall(ingName, expectedTags, expectedPorts)
```

which improves readibility by quite a bit, since I can just assume that the functions are doing what their names suggest, and understand the high level functionality we're testing by reading the `It()` block

MrHohn replied:
Seems like Github hides the content of the new file `test/e2e/firewall.go`. Need to click on `Load diff` to see it...

bprashanth commented on test/e2e/ingress.go:
```diff
@@ -135,6 +143,32 @@ var _ = framework.KubeDescribe("Loadbalancing: L7 [Feature:Ingress]", func() {
+			// Retrieve nodeTags from the node instance as the target tags for the ingress firewall rule
+			nodes := framework.GetReadySchedulableNodesOrDie(jig.client)
+			Expect(len(nodes.Items) > 0).Should(BeTrue())
+			nodeTags, err := getInstanceTags(cloudConfig, nodes.Items[0].Name)
```
as a general rule for this file:     

* the jig encapsulates any Kubernets/talking-to-apiserver logic, so looking up the service, parsing out ports etc. 
* the specific controller (be it gce or nginx) encapsulates gce or nginx specific logic (eg: listing firewalls etc)

So I suggest partitioning all the naked functions you have into one of those two buckets as appropriate

MrHohn replied:
Implemented as functions of testJig and GCEIngressController.

bprashanth commented on test/e2e/ingress.go:
```diff
@@ -119,6 +125,8 @@ var _ = framework.KubeDescribe("Loadbalancing: L7 [Feature:Ingress]", func() {
 		})
 
 		It("shoud create ingress with given static-ip ", func() {
+			cloudConfig := framework.TestContext.CloudConfig
```
this should be encapsulated in the gceController struct https://github.com/kubernetes/kubernetes/blob/master/test/e2e/ingress_utils.go#L910

MrHohn replied:
Fixed

bprashanth reviewed (commented):
sorry for the nits, the ingress e2e is kind of hard to debug because it verifies a lot under the covers, so I think we need to pay attention to readability there

MrHohn commented:
Addressed comments and re-pushed. Firewall related util functions are moved into framework/firewall_util.go. Also put some service or ingress related functions back to service.go and ingress_utils.go.

I believe we are close this time.

MrHohn commented:
@k8s-bot verify test this, issue: #38044

MrHohn commented:
@k8s-bot cvm gce e2e test this

MrHohn commented:
@k8s-bot gce etcd3 e2e test this, issue: #27673

MrHohn commented:
Rebased.

MrHohn commented:
@k8s-bot gci gke e2e test this

MrHohn commented:
Rebased.

freehan commented on test/e2e/ingress_utils.go:
```diff
@@ -862,6 +881,50 @@ func (j *testJig) curlServiceNodePort(ns, name string, port int) {
+		Expect(err).NotTo(HaveOccurred())
+		nodePorts = append(nodePorts, strconv.Itoa(int(svc.Spec.Ports[0].NodePort)))
+	}
+	return nodePorts
```
nit: s/constructFwForIngress/constructFirewallForIngress

freehan commented on test/e2e/framework/firewall_util.go:
```diff
@@ -0,0 +1,298 @@
+}
+
+// ConstructFwForLBSvc returns the expected GCE firewall rule for a loadbalancer type service
+func ConstructFwForLBSvc(svc *v1.Service, nodesTags []string) (*compute.Firewall, error) {
```
nit: s/ConstructFwForLBSvc/ConstructFirewallForLBService

freehan commented on test/e2e/framework/firewall_util.go:
```diff
@@ -0,0 +1,298 @@
+)
+
+// This should match the formatting of makeFirewallName() in pkg/cloudprovider/providers/gce/gce.go
+func MakeFirewallNameForLBSvc(name string) string {
```
nit: s/LBSvc/LBService

freehan commented on test/e2e/framework/firewall_util.go:
```diff
@@ -0,0 +1,298 @@
+	return fws
+}
+
+func PackProtocolsPortsFromFw(alloweds []*compute.FirewallAllowed) []string {
```
s/PackProtocolsPortsFromFw/PackProtocolsPortsFromFirewall

freehan commented on test/e2e/ingress_utils.go:
```diff
@@ -862,6 +881,50 @@ func (j *testJig) curlServiceNodePort(ns, name string, port int) {
+}
+
+// constructFwForIngress returns the expected GCE firewall rule for the ingress resource
+func (j *testJig) constructFwForIngress(gceController *GCEIngressController) *compute.Firewall {
```
nit: s/constructFwForIngress/constructFirewallForIngress

freehan commented:
only nits. LGTM overall

MrHohn commented:
Thanks, fixes nits and adds docs comments for public functions.

MrHohn commented:
@bprashanth @freehan Do we want to let this in before Holiday?

freehan commented:
Sorry. I missed this one.

MrHohn commented:
All builds failed because of #38154(renamed clientset) and #38638.

The later one introduced `${CLUSTER_NAME}` and removed `${NETWORK}` from some of the firewall rules. It makes running multiple e2e clusters within one GCE project possible. Made corresponding changes in the new commit. @freehan
